### PR TITLE
POC for custom args in modules - DO NOT MERGE

### PR DESCRIPTION
**Pull request by maxulysse**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: modules/nf-core/cnvkit/batch/main.nf [hunk 1]
```diff
@@ -27,7 +27,8 @@ process CNVKIT_BATCH {
     task.ext.when == null || task.ext.when
 
     script:
-    def args = task.ext.args ?: ''
+    def args = task.ext.args && task.ext.custom_args ? task.ext.args + ' ' + task.ext.custom_args :
```
Comment (FriederikeHanssen) [inline modules/nf-core/cnvkit/batch/main.nf:30]: ooooh, so users wouldn't need to specify the whole args line anymore when overwritting it?

**Comment by maxulysse** [comment]
using:
```
/// OVERWRITE ALL
process {
    withName: 'CNVKIT_BATCH' {
        ext.args          = '--foo'
    }
}
```
args is `--foo`

using:
```
/// ADD TO ARGS
process {
    withName: 'CNVKIT_BATCH' {
        ext.custom_args          = '--bar'
    }
}
```
args is now `--method hybrid --diagram --scatter --bar`

File: modules/nf-core/cnvkit/batch/main.nf (hunk 1 above)
Comment (maxulysse) [reply modules/nf-core/cnvkit/batch/main.nf:30]: yes, that's my idea, I wanted to do a POC so I can pitch to core and maintainers.

**Comment by maxulysse** [comment]
Also, with this possibility, I can use the underlying pipeline logic and just add to default params.
In that case, by default we have params.wes, that specify args for cnvkitbatch, if I don't use this params, my new args are now:
`--method wgs --diagram --scatter --bar`, so just adding `--bar` to my args, without having to copy over the whole logic in a custom config file.

**Comment by maxulysse** [comment]
tests are obviously failing due to `--foo` and `--bar` not being real args, but just used here for the POC

File: modules/nf-core/cnvkit/batch/main.nf (hunk 1 above)
Comment (jfy133) [reply modules/nf-core/cnvkit/batch/main.nf:30]: Looking pretty good to me!

File: modules/nf-core/cnvkit/batch/main.nf [hunk 2]
```diff
@@ -27,7 +27,8 @@ process CNVKIT_BATCH {
...
+        task.ext.args ?: ''
```
Comment (ewels) [inline modules/nf-core/cnvkit/batch/main.nf:31]: Super nice! I was wondering about this the other day actually, whilst explaining to someone that they had to figure out the default args to add something 😅 

I find the logic slightly difficult to follow. Would something like this work and be a bit easier to read? (untested)

```suggestion
    def args = task.ext.args ?: ''
    if(task.ext.custom_args) args += ' ' + task.ext.custom_args
```

Comment (maxulysse) [reply modules/nf-core/cnvkit/batch/main.nf:31]: oh yeah, so much better indeed

Comment (maxulysse) [reply modules/nf-core/cnvkit/batch/main.nf:31]: let me try locally

Comment (maxulysse) [reply modules/nf-core/cnvkit/batch/main.nf:31]: what about that, to keep the same logic for both?
```suggestion
    def args = task.ext.args ?: ''
        args += task.ext.custom_args ? ' ' + task.ext.custom_args : ''
```

Comment (maxulysse) [reply modules/nf-core/cnvkit/batch/main.nf:31]: > Super nice! I was wondering about this the other day actually, whilst explaining to someone that they had to figure out the default args to add something sweat_smile

I had some questions and requests about having this feature, and this morning as I woke up, I had the solution.
And for once, it was a good still a good idea when I finally try it out (not often the case when thinking about things while being half asleep)

Comment (ewels) [reply modules/nf-core/cnvkit/batch/main.nf]: Honestly, I find the ternary operator harder to read and I'm not sure it's needed here as we don't need to add an empty string if it's not set.

But I don't mind really, the main thing was to keep the logic separate for the two `task.ext` and you've kept that 👍🏻

Comment (asp8200) [reply modules/nf-core/cnvkit/batch/main.nf]: 🤯 I can't understand any of that. I general a prefer a lot of easy-to-understand lines as opposed to a few cryptic lines that I can't parse.

Comment (maxulysse) [reply modules/nf-core/cnvkit/batch/main.nf]: We have ternary operators everywhere. That's the third in this module alone. And I'm just copying the logic from the previous one

**Comment by mahesh-panchal** [comment]
So, just to put here what I put in the other thread, is that I'm not in favour of this. It means that mandatory non-file arguments are found in the `modules.config`, which if changed by a user could break the workflow. A user could be directed to copy over what's in `modules.config`, but this is not always fool proof as occasionally there are references to variables not defined in the `custom.config` but are in the `nextflow.config` ( default `params` for example - and I know the solution there is to just wrap it in a closure but the point remains) that allow it to work in `modules.config`. It also means defining another input `custom_args`. 

I really like that you coded up inheritance, but I think that `ext.args` should be the only method of defining args but additionally only the ones that don't break the workflow. If it wasn't for what can be in `ext.args`, I would say this is a good solution, since Nextflow configuration doesn't support inheritance. I feel like I'm raining on the parade here though.

My personal opinion is that users shouldn't be directed to `modules.config` to copy code they may not understand, but I guess this is actually another can of worms, because one needs to somehow communicate to the user which tool args they can override. Perhaps modules need a schema themselves, which would lead to being able to list configurable options for modules in pipelines, and a help indicating what string to write in `ext.args` ( rather than copying a potentially incomprehensible code block ), or better yet, a better way to configure tool args.

In the current state of things, I'm personally more in favour of pipeline developers patching their modules to maintain workflow stability, and then directing users to only supply strings to `ext.args`. For example, the workflow documentation explains that the flags X,Y,Z are controlled by the workflow, but the user can additionally supply, A, B, C, where B is currently the default, by doing ...

**Pull request closed without merging**